esi_leap/objects/lease.py

```diff
@@ -138,16 +167,41 @@ def _fulfill_lease():
         _fulfill_lease()
 
     def expire(self, context=None):
-        # unassign resource
-        resource = self.resource_object()
-        if resource.get_lease_uuid() == self.uuid:
-            resource.expire_lease(self)
+        leases = Lease.get_all(
+            {'parent_lease_uuid': self.uuid,
+             'status': [statuses.CREATED, statuses.ACTIVE]},
+            None)
+        for lease in leases:
+            lease.expire(context)
+        offers = offer_obj.Offer.get_all(
+            {'parent_lease_uuid': self.uuid,
+             'status': statuses.AVAILABLE},
+            None)
+        for offer in offers:
+            offer.expire(context)
+
+        @utils.synchronized(utils.get_resource_lock_name(self.resource_type,
+                                                         self.resource_uuid),
+                            external=True)
+        def _expire_lease():
```

Comment:
_expire_lease() and _cancel_lease() look similar. Can we take them out and make them one function?

Reply:
Potentially! That would fall out of the scope of this PR however.